core: mm: fix core virtual address range constraint in lpae

etienne-lms opened the pull request:
Removes constraint in LPAE implementation that expects core virtual
address range to fit a single block of the MMU base table used to
map the area. The implementation allows virtual addresses to spread
over several blocks of the base table. Therefor the limitation can be
removed.

As a result, selecting whether core virtual address range spreads
above or below the chosen base address depends on its location
against the virtual range mid address.

Fixes: https://github.com/OP-TEE/optee_os/issues/5201
Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -886,6 +886,11 @@ static void assign_mem_granularity(struct tee_mmap_region *memory_map)
 
+static bool place_tee_ram_at_top(paddr_t paddr)
+{
+	return paddr > core_mmu_get_va_width() / 2;
```
BIT64(core_mmu_get_va_width()) ?

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -813,14 +813,6 @@ void core_init_mmu(struct tee_mmap_region *mm)
-	size_t base_level_size = BASE_XLAT_BLOCK_SIZE;
-	paddr_t base_level_mask = base_level_size - 1;
-
-	return (paddr & base_level_mask) > (base_level_size / 2);
```
With this we were trying to avoid crossing boundaries of the top level tables, typically 1G on Armv7.
Crossing such a boundary costs 4k since we need to use one more translation table.

Perhaps this small memory saving isn't relevant on LPAE enabled platforms.

etienne-lms commented:
review comments addressed

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -1003,18 +1024,29 @@ static bool assign_mem_va(vaddr_t tee_ram_va,
+		/* Retry to setup mapping consumming an extra xlat table */
+		tee_ram_at_top = place_tee_ram_at_top(tee_ram_va);
+		retry = true;
+		goto define_vmap;
```
This is not how loops are implemented. Please find another way.

etienne-lms replied:
i agree it is not clean. i'll update...

etienne-lms replied:
I've split `assign_mem_va()` in 2 functions to ease integration.

jenswi-linaro reviewed (commented):
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented:
squashed fixup commits and applied review tag.
I rebased for CI dependency purpose.

Note I rephrased the commit message since original was not accurate.
